generalize async notification and add queued operation manager

pabigot opened the pull request:
SUPERSEDED by #23333

Addressing #22494 in support of #21538 this PR:
* extracts the async notification infrastructure from the on-off service into a self-contained structure that can be used anywhere asynchronous operations may be used;
* recasts the on-off service as a manager that is to be used by services;
* adds a manager for queued operations that handles all the queue and event processing generically, supported by service-specific helper functions for validation, processing, and completion notification.

See documentation ~at https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/22853/docs/reference/index.html~ by digging into the latest results of the Documentation Github Action, finding the artifact, downloading it, uncompressing this, untarring that, and pointing your browser at the html.

nordic-krch committed:
sys: onoff: Move transition functions out of service struct

Extracted transition functions from onoff structure to external one
which allows to keep them in flash.
Added ONOFF_SERVICE_DEFINE macro for static initialization.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

pabigot committed:
lib: queued_operation: add API for managing a queue of operations

Full asynchronous support for APIs such as bus transactions generally
require managing operations from unrelated clients.  This API provides
a data structure and functions to manage those operations generically,
leaving the service to provide only the service-specific operation
description and implementation.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

nordic-krch commented on lib/os/queued_operation.c:
```diff
@@ -0,0 +1,197 @@
+	__ASSERT_NO_MSG(op != NULL);
+
+	if (mgr->current == op) {
+		mgr->finalizing = true;
```
what is the purpose of `finalizing` flag? Is it to allow submitting new operation with higher priority from the callback?

pabigot replied:
Yes, it ensures that any changes to the queued operations performed during the callback affect the selection of the next operation to run.

nordic-krch commented on lib/os/queued_operation.c:
```diff
@@ -0,0 +1,197 @@
+	k_spin_unlock(&mgr->lock, key);
+
+	if (rv == 0) {
+		queued_operation_finalize(mgr, op, -ECANCELED);
```
why not just call 
```
async_notify_generic_callback cb = async_notify_finalize(&op->notify, res);

if (cb != NULL) {
	mgr->vtable->callback(mgr, op, cb);
}
```

now, if it interrupts `queued_operation_finalize` called from other context which will set finalize to true then this will clear it prematurely.

pabigot replied:
Good catch; fixed.

Thanks.  All documentation comments in this review have been addressed.  Please resolve what you think is done, and provide more information if you think there's something that wasn't addressed.

nordic-krch reviewed (commented):
few comments. I will continue tomorrow.

pabigot commented on include/sys/async_notify.h:
```diff
@@ -0,0 +1,345 @@
+
+	*notify = (struct async_notify){
+#ifdef CONFIG_POLL
+		.method = {
```
Without the `#ifdef` the header won't compile when `CONFIG_POLL=n` because it references a field that isn't defined in that case.  It should be possible to use async notify without enabling poll, so in this case the problem would be diagnosed at runtime by refusing to accept the request.

nordic-krch commented on lib/os/queued_operation.c:
```diff
@@ -0,0 +1,204 @@
+	if (rv < 0) {
+		return rv;
+	}
+
```
probably in majority of cases there will single operation only, it would be good to optimized this. I've been thinking about two things:
- using only `current` to indicate that module is busy, finalizing stage can be encoded by using invalid address (e.g. `UINTPTR_MAX`)
- when it is used then process can be started immediately and locklessy (note that for that atomic.h must be extended: #22680 :
```
if (atomic_cas((atomic_t *)&mgr->current, (atomic_t)NULL, (atomic_t)op)) {
	mgr->vtable->process(mgr, op);
	return rv;
}
```

I tried it (and adapt getting next op) and measured time which went from 5.5us to 3.9us on nrf52.
I was trying if it is feasible (https://github.com/nordic-krch/zephyr/commit/47371da4ec69fec96ce93a7e3b06e5c440740379)

pabigot replied:
OK, I've addressed all the concerns except this one.  (Aside: I think you have the wrong link for an extension to atomic.h.)

I'm not taking this suggestion, as I think it adds significant complexity (magic non-pointer values stored in pointers; a new synchronization mechanism) with little real value (1.5 us per submitted operation).

nordic-krch replied:
> I'm not taking this suggestion, as I think it adds significant complexity 

Here i disagree because:
- one variable to keep busy state which is simpler
- selecting next to process is simplified
- speed up (you say 1.5us, I say 30%)
- a bit less rom (150 bytes less)

nordic-krch commented on include/sys/queued_operation.h:
```diff
@@ -0,0 +1,243 @@
+ * function have been completed.
+ *
+ * During the call to this function the service process function will
+ * have been invoked at least once, either providing another operation
```
I still not sure if this statement is true. If manager is busy then new submit will just put operation into the queue and process function will not be called.

nordic-krch commented on include/sys/queued_operation.h:
```diff
@@ -0,0 +1,239 @@
+ * or indicating that no operations are pending.
+ *
+ * @param mgr a generic pointer to the service instance
+ * @param op a generic pointer to an operation to be performed
```
I think that it is worth mentioning that `notify` field in `struct queued_operation` must be initialized prior calling submit using async_notify_init api.

nordic-krch commented on include/sys/queued_operation.h:
```diff
@@ -0,0 +1,239 @@
+ *
+ * This function must be invoked by services that support queued
+ * operations when the operation provided to them through the process
+ * function have been completed.  It is not intended to be invoked by
```
maybe it's worth stressing that this should be called at the and of completion notification because it may result with process call with NULL operation before actual action is performed. For example if process function is synchronous then finalizing may look like:
```
void process(struct queued_operation_manager *mgr,
		struct queued_operation *op)
{
	if (op) {
		queued_operation_finalize(mgr, op, 0);
                // do some processing
	} else {
		// shut down resources
	}
}
```
In the case above process(mng, NULL) will be called in the context of `queued_operation_finalize` call, before actual action and that's probably unintended.

It is implicitly explained in the sentence "During the call..." but i wonder if it is enough.

nordic-krch commented:
I've been wondering about a way to use manager without priority, to avoid traversing whole list (in locked context) in each submit. E.g. use some reserved priority value which would mean just append to the list.

In the manager you don't provide functions for initializing async_notify maybe then onoff should deprecate those and also expect that user will call async_notiify directly.

I'm done with review. I like this api. What about splitting PR in 2 (async notify first). Manager is missing tests and will probably take longer to review.

pabigot commented:
You're right that `process()` needs to be documented as re-entrant (a term we haven't yet formally defined), and that if it's synchronous that could cause stack problems due to recursion.  I need to think about how to document that.  I don't really expect this to be used for synchronous operations, except in testing, because you don't need a queue for them.

> I've been wondering about a way to use manager without priority, to avoid traversing whole list (in locked context) in each submit. E.g. use some reserved priority value which would mean just append to the list.

I think that can be done, yes.
 
> In the manager you don't provide functions for initializing async_notify maybe then onoff should deprecate those and also expect that user will call async_notiify directly.

You may be right.  I don't want to deprecate anything right now, because I do want to rename `onoff_service` to `onoff_manager` for terminology reasons, and that requires more thought.

> I'm done with review. I like this api. What about splitting PR in 2 (async notify first). Manager is missing tests and will probably take longer to review.

I'm OK with splitting it, but don't understand the comment about missing tests.  The manager has 100% line-coverage in its test.

nordic-krch commented:
> but don't understand the comment about missing tests. The manager has 100% line-coverage in its test.

Were they in PR yesterday? I didn't see them yesterday, in fact they are included.

pabigot commented:
> Were they in PR yesterday? I didn't see them yesterday, in fact they are included.

They've been in the PR since it was first submitted.

nordic-krch commented:
> They've been in the PR since it was first submitted.

Never mind, i must have missed them somehow.

nashif commented on doc/reference/kernel/other/async_notify.rst:
```diff
@@ -0,0 +1,25 @@
+.. _async_notification:
+
```
I propose to move this out of kernel/other and put it directly under doc/reference under a new category, for example doc/reference/services or whatever.

pabigot committed:
lib: async_notify: add generic asynchronous notification infrastructure

k_poll() for a signal is often desired for notification of completion
of asynchronous operations, but there are APIs where it may be
necessary to invoke "asynchronous" operations from contexts where
sleep is disallowed, or before the kernel has been initialized.
Extract the general notification solution from the on-off service into
a utility that can be used for other APIs.

Also move documentation out to a resource management section.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Latest push has no change to implementation, just moves the documentation out of the kernel area.  Also rebased on current master.

anangl commented on include/sys/async_notify.h:
```diff
@@ -0,0 +1,346 @@
+ * are cleared by async_notify_validate().  Use of these does not
+ * imply that the flags field becomes public API.
+ */
+#define ASYNC_NOTIFY_EXTENSION_POS 3
```
Is the bit at position 2 reserved for some purpose?

anangl commented on include/sys/async_notify.h:
```diff
@@ -0,0 +1,346 @@
+ *
+ * These bits are intended for use by containing service
+ * implementations to record client-specific information.  The bits
+ * are cleared by async_notify_validate().  Use of these does not
```
`async_notify_validate` does not seem to clear these bits, they are just ignored there.

anangl commented on include/sys/async_notify.h:
```diff
@@ -0,0 +1,346 @@
+ * Mask isolating the bits of async_notify::flags that are available
+ * for extension.
+ */
+#define ASYNC_NOTIFY_EXTENSION_MASK (~(BIT(ASYNC_NOTIFY_EXTENSION_POS) - 1U))
```
Why not define it as `(~BIT_MASK(ASYNC_NOTIFY_EXTENSION_POS))`?

anangl commented on include/sys/async_notify.h:
```diff
@@ -0,0 +1,346 @@
+	 * Bits at and above ASYNC_NOTIFY_EXTENSION_POS are available
+	 * for use by service extensions while the containing object
+	 * is managed by the service.  They are not for client use,
+	 * and are cleared before completion notfication.
```
`async_notify_finalize` clears only the bits below ASYNC_NOTIFY_EXTENSION_POS. Was the intent to clear all of them?

anangl commented on include/sys/async_notify.h:
```diff
@@ -0,0 +1,346 @@
+ * notification configuration is valid and clears the result.
+ *
+ * Be aware that on return of success all extensible bits of the flags
+ * field will have been cleared; the caller must initialize them after
```
Currently `async_notify_validate` does not clear these bits.

pabigot commented:
@anangl thanks; the policy for handling service extensions using async_notify flags has been in flux.  A write to `flags` is used to release a spin wait, so IMO it's more clean to have finalization just store zero rather than attempt to preserve service-specific flags.  However the validation at this level can't do anything with them.

I've updated the documentation and the implementation to be more clear on who and when the content is available and preserved.

Also added a dedicated test for the async_notify infrastructure that verifies the flag clear/preservation behavior.

anangl commented on lib/os/queued_operation.c:
```diff
@@ -0,0 +1,216 @@
+	mgr->finalizing = false;
+
+	if (can_start(mgr)) {
+		select_next_and_unlock(mgr, key);
```
At this point it is already known that `mgr->current` is `NULL` and `mgr->finalizing` is `false`, but both `can_start` and `select_next_and_unlock` will check those (the checks from `can_start` will be probably optimized by the compiler).
I'd propose to refactor this a bit and do these checks only in `queued_operation_submit`, right before calling `select_next_and_unlock`, and here to call `select_next_and_unlock` directly.

pabigot replied:
I've just removed the extra check and called `select_next_and_unlock()` in both places.  I don't think there's any other refactoring that would be safe: `select_next_and_unlock()` needs the check to behave properly with operations that complete during the initial process call.

anangl commented on lib/os/onoff.c:
```diff
@@ -40,36 +39,11 @@ static int validate_args(const struct onoff_service *srv,
-		break;
-	default:
+	if ((rv == 0)
+	    && ((cli->notify.flags & ASYNC_NOTIFY_EXTENSION_MASK) != 0)) {
```
Is this actually needed? Why not just `return async_notify_validate(&cli->notify)` here?

pabigot replied:
This catches a situation where the client structure has not been initialized but the limited check of `async_notify_validate()` passes.  That requires only that the method be correct and consistent; it doesn't look for other bogus values.  This service doesn't currently use extension bits, so those should be zeroed if the structure was properly initialized.

A unit test fails if this check is removed: in that case the structure appears to be a valid callback configuration because the low two bits of flags identify callback and there's a non-null pointer, but in fact the pointer is garbage and a segfault occurs if its invoked.

pabigot commented on include/sys/queued_operation.h:
```diff
@@ -0,0 +1,280 @@
+ *
+ * @param priority the priority of the operation relative to other
+ * operations.  Numerically lower values are higher priority.  Values
+ * outside the range of a signed 8-bit integer will be rejected.
```
Thread priorities are signed integers; it seemed to make sense to have operation priorities be the same.

anangl replied:
Negative priority values may suggest some special treatment of operations that use those. For threads we've got preemptible and cooperative ones. Here we don't have any difference. That's why I thought that unsigned values seemed more natural in this case. But I won't insist on this.

anangl commented:
> I've updated the documentation and the implementation to be more clear on who and when the content is available and preserved.

Documentation still mentions that `async_notify_validate` clears the extensible flags (see https://github.com/zephyrproject-rtos/zephyr/pull/22853#discussion_r381330619 and https://github.com/zephyrproject-rtos/zephyr/pull/22853#discussion_r381340299), but in fact it does not touch them (and it should not IMO, I think the documentation should be corrected).

nordic-krch commented on lib/os/async_notify.c:
```diff
@@ -0,0 +1,73 @@
+	 * field, and notify if requested.
+	 */
+	notify->result = res;
+	notify->flags = 0;
```
maybe use ASYNC_NOTIFY_METHOD_COMPLETED here instead of 0

nordic-krch replied:
There is a chance for race condition here. Flags are cleared and result is written so if someone interrupts and fetchs result it will indicate completion. If user then reuses client and overwrite method, signal or callback it will misbehave since new data will be used below.

pabigot commented on lib/os/queued_operation.c:
```diff
@@ -0,0 +1,216 @@
+		in_idle = (op == NULL);
+
+		key = k_spin_lock(&mgr->lock);
+	} while (op != NULL);
```
If an operation is found (`op` is not null) and the call to `process()` finalizes the operation, then the service is available but has not been informed of this fact by re-invoking `process()` with a null pointer.  (There's a test that fails when the loop is not taken, showing this can happen.)

anangl replied:
This does not seem to be the right way of handling this case. The call to `process()` with a null pointer should be rather done in the context of `queued_operation_finalize()`. Currently this will not happen, because `in_idle` is `true` when `select_nex_and_unlock()` is called from `queued_operation_finalize()`. Consequently, `process()` is called with a null pointer in the case that you described, but not when `queued_operation_finalize()` is called asynchronously. For instance, the following test will fail:
```
static void test_async_processing(void)
{
	struct operation operation;

	reset_service();
	service.async = true;
	service.process_rv = 142;

	operation_init_spinwait(&operation);
	service_submit(&service, &operation, 0);
	service_finalize(&service, service.process_rv);
	zassert_false(service.active, "service not idled");
}
```

pabigot replied:
OK, I see that.  Logic updated so the idle notification is sent in post-finalization.  Thanks.

anangl replied:
And now you can get rid of the loop here and apply what I suggested in https://github.com/zephyrproject-rtos/zephyr/pull/22853#discussion_r381904466. See https://github.com/anangl/zephyr/commit/4d1021deb18bb9fe6cc174671156e458dd0e50b2. All existing tests will pass.

anangl replied:
I just realized that both the current implementation and what I proposed above will not properly handle the following case:
- `select_next_and_unlock()` called for a finalized operation is preempted right before it is to call `process()` with a null pointer, and the preempting code calls `queued_operation_submit`

This will result in `process()` being called to start the new operation and right after that this preempted call to `process()` (with a null pointer) will be done.

pabigot replied:
Yes, that makes sense.  We have to release the lock before calling process; I'll look into this later today.

pabigot replied:
I suspect the problem identified above is part of the reason I originally had this in a loop, though the original logic wasn't quite right either.

It's back in a loop: after process() has been invoked the manager needs to check for delayed finalizations and additions to the active operation queue while the manager was unlocked.  This is supported by a more robust four-state machine that can distinguish the notification phase (unlocked invocation of process()) from processing and finalizing.

I think this is right, but I haven't added a specific test case yet.  I won't get back to this for a while, but a review of the updated logic would be helpful.

anangl replied:
Yes, now I see the need for having a loop in `select_next_and_unlock()`. Hovewer, the logic used in this function can be made much simpler and thus easier to understand, see https://github.com/anangl/zephyr/commit/586b5674df42b8abd330add02694008616af89e4.
 Every iteration of the loop should result in a call to `process()`, either to announce a new operation or to signal that there is nothing more to do. Otherwise, if nothing new was added to the queue in the notification phase and the currently processed operation hasn't finalized in that phase, there is no need to execute the loop again. Or am I missing something?
Actually, the current definition of `notify`:
```
bool notify = (op != NULL) || (state != ST_IDLE);
```
makes it always `true`, as `op` can be `NULL` only when `state` is `ST_FINALIZING` (there's an assertion right above that checks this).

pabigot replied:
Agreed: that was a residue of a time before `lock_and_check_reselect()` handled both the logic and the state setting.

pabigot commented:
@anangl @nordic-krch thanks, updates pushed.

nordic-krch commented on lib/os/async_notify.c:
```diff
@@ -0,0 +1,80 @@
+	 * completed state, releasing any spin-waiters.
+	 */
+	compiler_barrier();
+	notify->flags = ASYNC_NOTIFY_METHOD_COMPLETED;
```
now, client ownership is not returned to the caller before signal notification happens. If higher priority thread polls, it will be waken up by signal raise before flag is cleared. I think that for signal case you need to have lock around flag cleaning and signal raise.

nordic-krch replied:
or just store signal to local variable, clear flags and raise signal?

pabigot replied:
Yes, the second is the path I took.  Good catch, thanks.

pabigot commented:
Rebased on top of #22419 to avoid future merge conflicts and simplify development targeting next release.

carlescufi commented on include/sys/async_notify.h:
```diff
@@ -0,0 +1,352 @@
+ * However, services that use this structure may access certain
+ * fields directly.
+ */
+struct async_notify {
```
In  general I think we are going way too far with name lengths in Zephyr.
I suggest we rename this to `anotify`,  here and  everywhere else  where `async_notify` appears.

jfischer-phytec-iot replied:
agree with @carlescufi, most of the names are too long:
`async_notify_generic_callback`
`async_notify_fetch_resul`
`async_notify_init_spinwait`
`ONOFF_SERVICE_TRANSITIONS_INITIALIZER`
`ASYNC_NOTIFY_METHOD_COMPLETED`
`ASYNC_NOTIFY_EXTENSION_MASK`

andrewboie replied:
I have a feeling this header is not namespacing stuff properly.
Shouldn't everything be prefixed with sys_?

carlescufi commented on include/sys/queued_operation.h:
```diff
@@ -0,0 +1,283 @@
+ * Instances of this should be members of a service-specific structure
+ * that provides the operation parameters.
+ */
+struct queued_operation {
```
Again, this is too long. I suggest `qdop` or  something similar.

pabigot commented:
In the absence of clear and consistent guidance naming seems to largely depend on personal taste I prefer names that unambiguously identify the capability, and (try to) use them consistently for the related data structures, functions, and constants.  If that's not the consensus primary consideration for Zephyr, I'll change it to whatever it's supposed to be.

Whether that includes adding the prefix `sys_`, and the implications of using that prefix, are also things people can discuss.  Many things accessed through `<sys/foo.h>` don't have that prefix, especially things that aren't closely tied to the kernel, which is the case with these APIs.

anangl commented on lib/os/queued_operation.c:
```diff
@@ -0,0 +1,206 @@
+	__ASSERT_NO_MSG(mgr != NULL);
+	__ASSERT_NO_MSG(op != NULL);
+
+	if (mgr->current == op) {
```
The function should not continue when these two pointers are not equal, i.e. an attempt is made to finalize an operation that is not currently processed.

pabigot replied:
That's intentional; operations that are cancelled are finalized to `-ECANCELED` but are not the current operation.  However, handling of the `finalize` field is incorrect in that case, in a way I thought I'd fixed....

nordic-krch replied:
but `queued_operation_finalize()` is no longer called from `queued_operation_cancel()` context. It is only API call now. I think in that case `struct queued_operation *op` should be removed from arguments list.

jfischer-phytec-iot commented:
> In the absence of clear and consistent guidance naming seems to largely depend on personal taste. I prefer names that unambiguously identify the capability, and (try to) use them consistently for the related data structures, functions, and constants. If that's not the consensus primary consideration for Zephyr, I'll change it to whatever it's supposed to be.

There is one guidance, `In general, follow the Linux kernel coding style...`
https://www.kernel.org/doc/html/v4.10/process/coding-style.html#naming
But few names here are too long, I think `callback` can be shorten to `cb` and as @carlescufi suggested `async_notify` to `anotify`. IMHO `sys_` is not necessary.

nordic-krch commented:
regarding names, I also think that they are ok. I'm for the rule: use abbreviations which are common otherwise full word. Unless you want to establish new common abbreviation (like dt in zephyr/linux world). `async` is already an abbreviation. As for callback i agree that this would be `cb` or `clbk` but i would keep `queued` and maybe consider `op` instead of operation since that is rather common abbreviation.

mbolivar commented on include/sys/queued_operation.h:
```diff
@@ -0,0 +1,278 @@
+/**
+ * @brief Helper to extract the result from a queued operation.
+ *
+ * This forwards to :cpp:func:`async_notify_fetch_result()`.
```
This isn't rendering properly. Looks like a missing `@rst ... @endrst`.

pabigot replied:
I've gotten rid of the RST syntax, fixed the typo, and changed to "like".  Please unresolve anything that isn't satisfactory or add a new comment.

mbolivar commented on include/sys/queued_operation.h:
```diff
@@ -0,0 +1,278 @@
+ * @brief Special priority value to indicate operation should be
+ * placed last in current queue.
+ *
+ * This is equvalent to providing the lowest priority but uses a
```
```suggestion
 * This is equivalent to providing the lowest priority but uses a
```

mbolivar commented on include/sys/queued_operation.h:
```diff
@@ -0,0 +1,278 @@
+ * @brief Special priority value to indicate operation should be
+ * placed first in the current queue.
+ *
+ * This is similar to providing the highest priority but uses a
```
As a first-time reader, the fact that this uses "similar" but the previous docstring uses "equivalent" makes me wonder if this is somehow not "equivalent" -- if not, why not? If it is equivalent, why not just use "equivalent" again?

pabigot replied:
The standard insertion uses FIFO within a priority.  When you append the effect is FIFO, which is equivalent.  When you prepend the effect is LIFO, which is not equivalent.

I assumed going into the details would be distracting, as the difference is implicit in the description.  I can change "equivalent/similar to" to "like" for both if that reads better.

mbolivar reviewed (commented):
A few drive by typos.

nordic-krch commented:
@pabigot i've been trying to go back to the review after couple of days and noticed complexity increase. I came to the conclusion that the reason for that is the loop in `select_next_and_unlock()` and loop is necessary to handle notification about no pending operations which contrary to other `process` calls is synchronous (no callback to call `queued_operation_finalize()`). It's goal is to notify about going into idle but imo it has limitations because it is synchronous. It transfers complexity to the module underneath because:
- module must maintain state to ramp up on first request after idle
- if going to idle is asynchronous (e.g. scheduling i2c transfer to go to low power mode) then module must support handling (queueing) request which comes during transition to idle so module underneath can no longer be simple executor of asynchronous operations.

Because of that i would like to propose following change in the api: add `setup(mgr, callback)` and `teardown(mgr, callback)`. Those functions are optional and since they only drive internal manager state the callback can be something like `callback(struct queued_operation_manager *mgr, int res)`.
With that change every operation in the manager is asynchronous thus managing will simplify significantly. Also because those are optional contrary to `process(mgr, NULL)` call.
Note that I started to work on asynchronous sequence manager where i used similar approach: https://github.com/nordic-krch/zephyr/blob/async_seq/include/sys/async_seq.h

pabigot commented:
It's not immediately obvious how adding two optional functions to the API simplifies anything, because there's a inherent complexity to notification/processing/finalization that was rediscovered as @anangl and I went through the previous comments.

However you raise an important technical point:  `process()` is async when given a non-null operation parameter, but the manager state machine requires it to be synchronous when being told no more work is available because the manager immediately transitions to `IDLE`.  In fact the startup and shutdown operations may be asynchronous which as you point out complicates the service implementation because it must delay with subsequent notifications while it's still in transition.  Which is completely contrary to the goal.

But the limitations you describe are exactly the sort of thing the onoff manager is intended to support.  So the best way to deal with this may be to add an optional onoff service pointer to the manager state.  (This is also an "eat your own dogfood" thing, because the set of manager capabilities were always intended to be components that could be composed to support specific services.)

If the queued operation manager does not expose an onoff service, then the manager considers transitions from `OFF` to `ON` to be handled internal to the service.  The existing manager state machine has to stay because it's what gets executed when the service is on, and `process()` is a straightforward async function that is invoked once for each operation to be processed, with no null pointer magic as an out-of-band indication (that's done by telling the service to turn off).

I think this has some promise, and will dig into it.  Thanks for raising this concern.

nordic-krch commented on include/sys/queued_operation.h:
```diff
@@ -0,0 +1,278 @@
+	 *
+	 * @param cb the generic callback to invoke.
+	 */
+	void (*callback)(struct queued_operation_manager *mgr,
```
shouldn't it contain result? I know that it is written inside async_notify in operation but it will definitely be needed and fetching it inside callback seems strange as it is already available when this callback is invoked.

pabigot replied:
I don't think it's necessary.  The intermediary translation function provided by the service can extract the result if passing it to the end user as a distinct argument is useful.  In a probably common case where the completed operation is simply appended to another list for processing in another context it isn't needed.

pabigot commented:
> So the best way to deal with this may be to add an optional onoff service pointer to the manager state.

I've reworked the state machine to use an optional onoff manager to handle the transitions between `OFF` (no operations pending) and `ON` when the service needs to be informed about them.  This involved some minor rework of the inner handling of success/error indicators.

The management is slightly simpler now for states specific to transitioning queued operations (`IDLE`, `NOTIFYING`, `PROCESSING`,`FINALIZING`).

Introduction of the onoff manager introduces the concept of an error in the queued operation manager.  There will have to be new API to detect and react to that; it'll come soon.

I plan to close this PR soon and split out the pieces related only to async notify and onoff manager to another PR which is more likely to be merged quickly.  A new PR in draft format based on that rework will be added to continue work on this feature.  Any comments on the changes made here will be addressed in the initial post of that draft PR.

nordic-krch commented on lib/os/queued_operation.c:
```diff
@@ -0,0 +1,444 @@
+{
+	bool loop = false;
+
+	do {
```
i still don't understand why you need a loop here. I only see a benefit that if process function is synchronous then process is not called recursively but this isn't a goal of this module. User is already warned in the header file that synchronous process may lead to recursive calls. We rather don't expect process to be synchronous because if it would be then there is no point of queuing operations.

Implementation is very complex now, with multiple places where interrupts are locked and unlocked in other functions. It's really hard to understand that. Especially that flow seems simple with only two forks: setup failure, new request pending after teardown.

pabigot replied:
The last time I incrementally refactored this based on a belief a loop wasn't necessary, it turned out that it was necessary.  So I'm going to prioritize correctness.  We can reconsider whether it's possible to avoid looping to react to events received during unlocked periods after the solution has been shown to work correctly and be functionally complete.

nordic-krch commented on lib/os/queued_operation.c:
```diff
@@ -0,0 +1,444 @@
+	ST_STOPPING,
+
+	/* Service is in an error state. */
+	ST_ERROR,
```
seems that there is no way to get out of error state. How to reinitialize the manager?

pabigot commented on include/sys/queued_operation.h:
```diff
@@ -0,0 +1,292 @@
+	/* Pointer to an on-off service supporting this service.  NULL
+	 * if service is always-on.
+	 */
+	struct onoff_service *onoff;
```
I'm somewhat sympathetic to that position, but I'm going to stick with this approach for now.

A driving goal of the manager framework is to provide components that can be composed to solve complex problems without re-implementing things and trying to handle all the edge cases (e.g., invocation from within an interrupt handler).  The onoff manager already provides all the necessary logic to safely and correctly handle an asynchronous transition between an off and on state, regardless of calling context and current underlying state.  If it does something wrong, it needs to be fixed.  Problems are more likely to be found and addressed in the onoff manager component, than in edge cases that don't get exercised by a specific queued operation manager callback solution.

Also, at least in the near term, the underlying services used by a queued operation manager (e.g. I2C or SPI) may continue to be used without a queued operation manager.  They may also need to be able to be turned on and off as we start to inch towards functional power management.  If that capability is provided through the onoff manager abstraction then it's available for both ways of using the service.

pabigot commented:
Superseded by #23333.
